Introduce the SKB metadata extern

vbnogueira opened the pull request:
We are adding the SKB metadata extern to the tc backend's control blocks and deparser definitions
The main difference from the previous version is that we dropped the extern representation of the
skb( `tc_skb_metadata`), and made skb read/write operation into extern functions.
So, for example,  to set the skb protocol, the user should do the following:

```
skb_set_protocol(my_new_proto);
skb_set_meta();
```

And to retrive the skb protocol:

```
skb_get_meta();
proto = skb_get_protocol();
```

vbnogueira commented:
@komaljai please take a look when you have some time

komaljai commented on backends/ebpf/psa/ebpfPsaControl.h:
```diff
@@ -71,7 +74,9 @@ class EBPFControlPSA : public EBPFControl {
     EBPFControlPSA(const EBPFProgram *program, const IR::ControlBlock *control,
                    const IR::Parameter *parserHeaders)
-        : EBPFControl(program, control, parserHeaders) {}
+        : EBPFControl(program, control, parserHeaders),
```
can you add the skb_metadata in EBPFControlPNA class which is in TC

vbnogueira replied:
I'll pass this on to mouse

vbnogueira replied:
Was able to pass it to `EBPFControlPNA`

komaljai commented:
You can try to inherit EBPF class DeparserBodyTranslatorPSA in TC and make related changes in TC backend.

komaljai commented:
Could you add example for skb metadata extern?

vbnogueira commented:
> Could you add example for skb metadata extern?

I added the example in this commit https://github.com/p4lang/p4c/pull/4916/commits/96707ead39c624ec2a01dcdc6e80b2ff5e43b9f4 and rebased

vbnogueira commented:
> You can try to inherit EBPF class DeparserBodyTranslatorPSA in TC and make related changes in TC backend.

Done

vbnogueira commented:
Tested with other examples we had and saw that this PR breaks one of our examples
So converted this back to a draft PR while I figure out the solution

vbnogueira commented:
@komaljai please take another look when you have some time

komaljai commented on testdata/p4tc_samples_outputs/add_entry_1_example_control_blocks.c:
```diff
@@ -161,6 +178,12 @@ if (/* hdr->ipv4.isValid() */
         }
+
+        if (have_saved_proto) {
+            bpf_p4tc_skb_set_protocol(skb, &sa->set, saved_proto);
```
Is setting the protocol and skb_meta required everytime or can we optimize this and set only when skb externs are called

vbnogueira replied:
The setting of the protocol is a way to avoid `bpf_skb_adjust_room`'s limitation of only supporting IPv4 and IPv6
We want to avoid making the user aware of this because this would hurt usability and require them to know specifics about `bpf_skb_adjust_room` that shouldn't matter to them
So we chose to do this automatically without involving the user
